drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-kogito-client/src/main/java/org/drools/workbench/screens/scenariosimulation/kogito/client/dmn/DMNTypeFactory.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.drools.workbench.screens.scenariosimulation.kogito.client.dmn;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import jsinterop.base.Js;
+import org.drools.workbench.screens.scenariosimulation.kogito.client.dmn.feel.BuiltInType;
+import org.kie.workbench.common.dmn.webapp.kogito.marshaller.js.model.dmn12.JSITItemDefinition;
+
+/**
+ * Utility factory for <code>DMNType</code>
+ */
+public class DMNTypeFactory {
+
+    private DMNTypeFactory() {
+        // Utility class, not instantiable
+    }
+
+    public static ClientDMNType getDMNType(final JSITItemDefinition itemDefinition,
+                                           final String namespace,
+                                           final Map<String, ClientDMNType> dmnTypesMap) {
+        List<JSITItemDefinition> jsitItemDefinitions = itemDefinition.getItemComponent();
+        if (jsitItemDefinitions != null && !jsitItemDefinitions.isEmpty()) {
+            Map<String, ClientDMNType> fields = new HashMap<>();
+            for (int i = 0; i < jsitItemDefinitions.size(); i++) {
+                final JSITItemDefinition jsitItemDefinition = Js.uncheckedCast(jsitItemDefinitions.get(i));
+                final String typeRef = jsitItemDefinition.getTypeRef();
+                if (!dmnTypesMap.containsKey(typeRef)) {
+                    ClientDMNType nestedClientDMNType = getDMNType(jsitItemDefinition, namespace, dmnTypesMap);
+                    dmnTypesMap.put(typeRef, nestedClientDMNType);
+                }
+                fields.put(jsitItemDefinition.getName(), dmnTypesMap.get(typeRef));
+            }
+            return new ClientDMNType(namespace, itemDefinition.getName(), itemDefinition.getId(), itemDefinition.getIsCollection(), true, fields, null);
+        } else {
+            return new ClientDMNType(namespace, itemDefinition.getName(), itemDefinition.getId(), itemDefinition.getIsCollection(), null);
+        }
+    }
+
+
+    public static class ClientDMNType {
```

Comment:
I think we can move it outside this class because I expect it will be used by many classes. Wdyt?

Reply:
@danielezonca Yeah, done